sql/core/src/main/scala/org/apache/spark/sql/execution/python/PythonArrowOutput.scala

```diff
@@ -211,15 +230,72 @@ class SliceArrowOutputProcessorImpl(
     batch
   }
 
-  override def loadBatch(): Boolean = {
-    if (rowCount > 0 && currentRowIdx < rowCount) {
-      true
+  protected override def getVectors(root: VectorSchemaRoot): Array[ColumnVector] = {
+    root.getFieldVectors.asScala.map { vector =>
+      new ArrowColumnVector(vector)
+    }.toArray[ColumnVector]
+  }
+
+  override def close(): Unit = {
+    if (prevRoot != null) {
+      prevVectors.foreach(_.close())
+      prevRoot.close()
+    }
+  }
+}
+
+class SliceBytesArrowOutputProcessorImpl(
+    reader: ArrowStreamReader,
+    pythonMetrics: Map[String, SQLMetric],
+    arrowMaxBytesPerOutputBatch: Int)
+  extends BaseSliceArrowOutputProcessor(reader, pythonMetrics) {
+
+  protected override def getRoot: VectorSchemaRoot = {
+    if (getBatchBytes(root) < arrowMaxBytesPerOutputBatch) {
+      currentRowIdx = currentRowIdx + root.getRowCount
+      root
     } else {
-      val loaded = super.loadBatch()
-      currentRowIdx = 0
-      loaded
+      var rootSlice = root
+      val stepRatio = 0.9
+      var sliceBatchCount = ((rowCount - currentRowIdx) * stepRatio).toInt
```

Comment:
Maybe, shall we add a comment about this formula and the magic number `stepRatio=0.9`?

Reply:
Yes, let me add a comment.